core/src/main/java/io/grpc/transport/MessageDeframer.java

```diff
@@ -124,27 +148,131 @@ public MessageDeframer(Listener listener) {
+    transportCount.addAndGet(numMessages);
+    requestedMessageCount += numMessages;
+    if (messagesAvailable && !appInRequest) {
+      appInRequest = true;
+      try {
+        listener.messagesAvailable(this);
+      } finally {
+        appInRequest = false;
+      }
+    }
+  }
+
+  /**
+   * Called in the application thread to consume the available messages provided by the transport.
+   * @param consumer which receives all available messages.
+   */
+  @Override
+  public int drainTo(StreamListener.MessageConsumer consumer) {
+    int count = 0;
+    do {
+      if (appQueue == null) {
+        appQueue = new ArrayDeque<Frame>();
+      }
+      if (appQueue.isEmpty()) {
+        // Get the queue from the transport and copy it into the appQueue
+        ArrayDeque<Frame> transportQueue;
+        do {
+          transportQueue = queueRef.getAndSet(null);
+          if (transportQueue != null) {
+            appQueue.addAll(transportQueue);
+            transportQueue.clear();
+          }
+          // Return it to the transport thread
+        } while (!queueRef.compareAndSet(null, transportQueue));
+        messagesAvailable = !appQueue.isEmpty();
+      } else {
+        // Consume the messages we already have. Will check for messagesAvailable
+        count += drainTo(consumer, appQueue);
+      }
+    } while (requestedMessageCount > 0 && messagesAvailable);
+    return count;
+  }
+
+  private int drainTo(StreamListener.MessageConsumer consumer, ArrayDeque<Frame> queue) {
+    int count = 0;
+    int consumedBytes = 0;
+    boolean endOfStream = false;
+    try {
+      Frame frame;
+      do {
+        frame = queue.peek();
+        if (frame == null) {
+          // Nothing to read
+          break;
+        }
+        if (frame == END_OF_STREAM) {
+          // Transport enqueued EOS, so empty queue and report to the listener
+          queue.clear();
+          endOfStream = true;
+          break;
+        }
+        // Can't deliver messages to the consumer or return flow control
+        if (requestedMessageCount == 0) {
+          break;
+        }
+        // Consume the message from the queue
+        queue.poll();
+        consumedBytes += frame.consumedBytes;
+        if (frame.stream != null) {
+          try {
+            // Calls to consumer may trigger a call to request in the same thread, we prevent
+            // the unnecessary call to notify again as this loop will pick up the additional
+            // count.
+            appInRequest = true;
+            count++;
+            consumer.accept(frame.stream);
+          } catch (Throwable t) {
+            t.printStackTrace();
+            // Decision point: Log / exit / drain the available messages
+          } finally {
+            try {
+              frame.stream.close();
+            } catch (IOException ioe) {
+              ioe.printStackTrace();
+              // Decision point: Worth logging?
```

Comment:
If there is already another exception happening, its probably not worth logging.  At that point this would be a suppressed message in java8.  Only logging when there isn't already another exception happening would allow clients to grab a copy of the logger and mute it if they wanted.